---
output: word_document
---
# Reanalysis the BRCA data

The RAW bam data download from GDC is extracted to subbam for recheck. And the calling process was done through GATK. Ignoring the `hard filter` provided by the GATK. Now I walk through the whole process for the project.

The outline follows the instruction:  
preprocessing:

1. Download BRCA WXS data from [GDC]('https://gdc-portal.nci.nih.gov/projects/TCGA-BRCA').
2. Dwonload Regulatory Build from Ensembl (release-v86).
3. Call mutation at regulatory regions by GATK (**-L regulatoryRegion.bed**)

Filteration: 

1. R script for obtain somatic mutation for every sample.
  + `02.getSomatic.R`
  + manifest statistics for file size and mutation.
  + **Normal: MQ > 20, DP >= 5, altDP >= 3**; write to 02.filtersnp
  + **Tumor: MQ > 20, DP >= 10, altDP >= 3**; write to 02.filtersnp
  + **Somatic: $Tumor -  Normal\ per\ case$**; write to 03.somatic/01.annotation
2. Bash script ANNOVAR to **filter dbSNP**
  + `removedbSNP.sh`
3. Annotate with **regulatory region**.
  + `regulatoryFeture.py`
4. Transform file to rds for analysis
  + `03.loadAnnoSomatic.R`
5. Reduce false positive site
  + `04.somaticMutationAnalysis.R`
  + **normalMutation recurrent >=1**
  + **$truePositiveSomaticMutation = somaticMutation - normalMutation$**
  + draw **01.SomaticMutationStat.png**
  + **Filter position which recurrent >= 5**
6. Refined recurrent mutation.
  + 37 mutation positions recur >=5 and save them to scan all mpileups.
  + `extractCoverage.sh` extract position from all samples. 
  + The process is working for the ratio and ensure the position is a real somatic mutation.
  + **Mutation point has count >= 10 and alt_count >=3 in normal**
  + **filter out normal false mtuation get depth >= 10 normal sample**
  + **02.depth_for_tumor_normal_barplot.png**
  + **03.mutated_samples_barplot.png**
7. Use biomaRt to get target genes for mutation
  + `05.targetGeneForMutation.R` get nearest target genes.
  + mark as Oncogene or tumor supressor genes.
8. Target gene expression
  + `06.targetGeneExpression.R` calculate gene expression.
9. Recheck the mutation point.
  + Get mpileup coverage `extractCoverage.sh`
  + Load point mpileup file for recheck, the position cov don't attached with base quality.
10. Do survival analysis of 2 candidate mutation
  + According to the mutation target gene expression, filtering out the candidate mutation
  + Use `08.survival_analysis.R` to do survival analysis
  + group cut by the gene epxression **high** **low**
  + group cut by the mutated gene **WT** and **MT**
11. Check and BLAT the two mutations
  + the perl code `Blat_alignedreads_onlychr_scorelimit_comparediscard_sanger.pl` rescale the ref and alt coverage through the mpileup and bam files.
  + the snp called from bam is based the on the reads mapped to the point, but the mapping allows **3 mismatches in 35 bp**. the snp can be mapping error or real snp. If the 
  + the two points are not located in the repeats.


Mutation in regulatory function.
1. Check mutation point in the regulatory region
  + check the point mutation in the regulatory region and motif.
  + check Wei Liu data to find TF binding region.
  + Check 


 
 
 